### DISCUSSION TITLE
How to substitute noble-ed25519 for node:crypto?

### DISCUSSION THREAD

**Discussion by guest271314** [paulmillr/noble-ed25519#98] [Q&A] [answered]
I'm currently using this from https://github.com/GoogleChromeLabs/webbundle-plugins/tree/main/packages/rollup-plugin-webbundle#isolated-web-app-signed-web-bundle

```
integrityBlockSign: {
  strategy: new wbnSign.NodeCryptoSigningStrategy(key),
},
```

There is supposed to be an option to do this

```
const privateKeyFile = "ed25519key.pem";
const privateKey = fs.readFileSync(privateKeyFile);

const build = async () => {
  const key = wbnSign.parsePemKey(
    privateKey,
  );

  const bundle = await rollup.rollup({
    input: "./src/script.js",
    output: { dir: "dist", format: "esm" },
    plugins: [
      wbnOutputPlugin({
        baseURL: new wbnSign.WebBundleId(
          key,
        ).serializeWithIsolatedWebAppOrigin(),
        static: { dir: "assets" },
        output: "signed.swbn",
        integrityBlockSign: {
          strategy: new (class {
            async sign(data) {
              return await ed.signAsync(data, key);
              // E.g. connect to one's external signing service that signs the payload.
            }
            async getPublicKey() {
              /** E.g. connect to one's external signing service that returns the public
               * key. */
              return await ed.getPublicKeyAsync(key);
            }
          })(), // new wbnSign.NodeCryptoSigningStrategy(key),
        }
```

which throws

```
node --experimental-default-type=module --experimental-network-imports rollup.wbn.js
(node:34843) ExperimentalWarning: Network Imports is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Error: Uint8Array expected
    at err (https://unpkg.com/@noble/ed25519@2.0.0/index.js:11:33)
    at au8 (https://unpkg.com/@noble/ed25519@2.0.0/index.js:15:5)
    at toU8 (https://unpkg.com/@noble/ed25519@2.0.0/index.js:17:26)
    at getExtendedPublicKeyAsync (https://unpkg.com/@noble/ed25519@2.0.0/index.js:247:53)
    at Module.getPublicKeyAsync (https://unpkg.com/@noble/ed25519@2.0.0/index.js:249:37)
    at rollup.rollup.plugins.wbnOutputPlugin.integrityBlockSign.strategy.getPublicKey (file:///home/user/webbundle/rollup.wbn.js:35:31)
    at Object.refinement (file:///home/user/webbundle/lib/index.js:245:60)
    at executeRefinement (file:///home/user/webbundle/node_modules/zod/lib/index.mjs:3468:39)
    at file:///home/user/webbundle/node_modules/zod/lib/index.mjs:3499:28
    at async file:///home/user/webbundle/node_modules/zod/lib/index.mjs:2560:29 {
  code: 'PLUGIN_ERROR',
  plugin: 'wbn-output-plugin',
  hook: 'generateBundle'
}
```

**Comment by paulmillr** [comment]
noble expects Uint8Array key and you're passing something else. Investigate how to convert the key to Uint8Array.

**Reply by guest271314** [reply]
I'v tried  few ways. I think the package is filled with `node:crypto` usage the still uses `node:crypto` on the user-defined object.

Let's use `ed` throughout to see what happens

```
node --experimental-default-type=module --experimental-network-imports rollup.wbn.js
(node:36852) ExperimentalWarning: Network Imports is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:36852) ExperimentalWarning: The Ed25519 Web Crypto API algorithm is an experimental feature and might change at any time
{
  key: Uint8Array(32) [
     18, 122, 118,  58, 109,  46, 221, 113,
    111,  63,  87, 197,  69, 221, 158, 168,
     24,  12,  26,  42, 137, 147, 128, 116,
     26, 239,  19, 177, 132,  51, 176,  31
  ],
  ed25519_privateKey: CryptoKey {
    type: 'private',
    extractable: true,
    algorithm: { name: 'Ed25519' },
    usages: [ 'sign' ]
  },
  node_key: PrivateKeyObject [KeyObject] { [Symbol(kKeyType)]: 'private' }
}
Error: WebBundleId: Only ed25519 keys are currently supported. Your key's type is undefined.
    at new WebBundleId (file:///home/user/webbundle/node_modules/wbn-sign/lib/web-bundle-id.js:14:19)
    at build (file:///home/user/webbundle/rollup.wbn.js:31:18)
    at file:///home/user/webbundle/rollup.wbn.js:70:1
```

I'm not sure if the requirement is possible without a complete re-write of https://www.npmjs.com/package/wbn-sign.

At `/lib/signers/signing-strategy-interface.d.ts` there is

```
/// <reference types="node" />
import { KeyObject } from 'crypto';
export interface ISigningStrategy {
    sign(data: Uint8Array): Promise<Uint8Array>;
    getPublicKey(): Promise<KeyObject>;
}
```

Web Cryptography API has `CryptoKey` objects, not `PrivateKeyObject` and `KeyObject` objects.

```
import * as fs from "node:fs";
import * as path from "node:path";
import { webcrypto } from "node:crypto";
import * as ed from "https://unpkg.com/@noble/ed25519";

const ed25519_key = await crypto.subtle.generateKey(
  "Ed25519",
  true, /* extractable */
  ["sign", "verify"],
);

// const crypto.subtle.exportKey('spki', x25519_key.publicKey);

const privateKeyFile = "ed25519key.pem";
const privateKey = fs.readFileSync(privateKeyFile);

const build = async () => {
  const node_key = wbnSign.parsePemKey(
    privateKey,
  );
  const key = ed.utils.randomPrivateKey();
  console.log({ key, ed25519_privateKey: ed25519_key.privateKey, node_key });
// ...
```

```
{
  key: Uint8Array(32) [
     44, 100,  72, 123,  53, 144, 230, 122,
    143,  82, 129, 120, 100, 207, 130, 184,
    149, 117,  61, 200, 187,  54,  55,  16,
     43, 197, 110,   3, 251, 118,  92, 220
  ],
  ed25519_privateKey: CryptoKey {
    type: 'private',
    extractable: true,
    algorithm: { name: 'Ed25519' },
    usages: [ 'sign' ]
  },
  node_key: PrivateKeyObject [KeyObject] { [Symbol(kKeyType)]: 'private' }
}
```

```
Error: WebBundleId: Only ed25519 keys are currently supported. Your key's type is undefined.
    at new WebBundleId (file:///home/user/webbundle/node_modules/wbn-sign/lib/web-bundle-id.js:14:19)
    at build (file:///home/user/webbundle/rollup.wbn.js:28:18)
    at file:///home/user/webbundle/rollup.wbn.js:67:1
```

https://github.com/nodejs/node/pull/38668#issuecomment-840704923

> Raw X/Ed key import was the reason why @jasnell implemented this node-specific webcrypto extension in the first place.

https://github.com/nodejs/node/issues/38655#issuecomment-840420467
> @vlovich it is worth pointing out that Node.js does not implement the proposal, it is intentionally using Vendor-specific proprietary extensions for Ed25519/X25519/Ed448/X25519. Nevertheless it was not intended to have them exportable, I've raised #38668 to correct this.

This is the source code I'm using https://github.com/guest271314/webbundle.

**Reply by guest271314** [reply]
This appears to indicate Node.js implementation is unique https://github.com/jacobbubu/webcrypto-ed25519/tree/main?tab=readme-ov-file#algorithm-identifier.

**Reply by paulmillr** [reply, marked as answer]
doesn’t seem to be a problem of noble.

we have very simple api: give u8a, take u8a. webcrypto uses custom objects, but they can be exported and imported to u8a

**Reply by guest271314** [reply]
Correct, it's probably not a problem of noble. I'm asking how to go about substituting noble code for Node.js code in a NPM package. I'm not an expert in Web Cryptography API. I don't know where to start substituting all of the Node.js-specific `node:crypto` code for either noble or Web Cryptography API code, which is the ultimate goal to create the Signed Web Bundle and Isolated Web App in the browser. Chromium 122 supports Ed25519. We can read and write files in the browser. I just need some help with the cryptography signing parts. I'll keep hacking away at what I'm trying to do either way. Thanks.

**Reply by paulmillr** [reply]
For interop with node or webcrypto, it’s simple: just use their export and import methods. 

the problem lies deeper than node api: the package in question wants to consume some specific key format.

**Reply by guest271314** [reply]
I don't think it's that simple. While the README says you can supply your own signing and get public key method, when you do it's based on Node.js `crypto` being involved. 

I think all Node.js `crypto` parts will have to be replaced for this code to run in the browser. That means going line by line to replace, because I don't think a tool exists that can do that.

If you think it's simple, kindly demonstrate how you would use your library to do that, as I am getting errors.

**Reply by paulmillr** [reply]
All node parts have to be replaced for browser - that’s correct

**Reply by guest271314** [reply]
Supposedly, since Ed25519 is now implemented in browsers we should only have to use `wbn-sign` and `wbn` packages 

https://github.com/GoogleChromeLabs/webbundle-plugins/issues/11#issuecomment-1847301072
> It's not limited to node-only as there's also golang version of the package. Whenever the support for ed25519 keys is added to [SubtleCrypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/sign#algorithm), it'll be rather simple to create a browser-only version of the [wbn-sign package](https://www.npmjs.com/package/wbn-sign). Web bundles are still experimental and I'm sure the tooling coverage improves when use of IWAs increases.

But nobody has answered the question as to details https://github.com/GoogleChromeLabs/telnet-client/issues/27.

Again, I am not a Web Cryptography API expert. I test and experiment on browsers and JavaScript runtimes, and mainly started experimenting with Signed Web Bundles because Chromium authors decided to gate Direct Sockets (`TCPSocket`, `TCPServerSocket`, `UDPSocket` in the browser) behind Isolated Web Apps, see https://github.com/guest271314/telnet-client. So I'm learning about this cryptography thing as I go along.

**Reply by guest271314** [reply]
FWIW I figured it out. Yesterday I decided to stop asking questions and for help and got to work. Replaced all the Node-js specific `crypto` references with `webcrypto` equivalents; along with `Uint8Array` for the `Buffer` usage. The same code can now be run by `node`, `deno`, and `bun` to produce the same signed web bundle and isolated web app. Thanks, again.
